hack/generated/controllers/suite_test.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/Azure/k8s-infra/hack/generated/controllers"
+	"github.com/Azure/k8s-infra/hack/generated/pkg/testcommon"
+)
+
+const (
+	TestRegion    = "westus"
+	TestNamespace = "k8s-infra-test-ns"
+)
+
+type ControllerTestContext struct {
+	*testcommon.TestContext
+	SharedResourceGroup *resources.ResourceGroup
+}
+
+func setup() (ControllerTestContext, error) {
+	ctx := context.Background()
+	log.Println("Running test setup")
+
+	SetDefaultEventuallyTimeout(2 * time.Minute)
+	SetDefaultEventuallyPollingInterval(5 * time.Second)
+
+	testContext, err := testcommon.NewTestContext(
+		TestRegion,
+		TestNamespace,
+		controllers.ResourceStateAnnotation,
+		controllers.ResourceErrorAnnotation)
+
+	if err != nil {
+		return ControllerTestContext{}, err
+	}
+
+	err = testContext.CreateTestNamespace()
+	if err != nil {
+		return ControllerTestContext{}, err
+	}
+
+	// Create a shared resource group, for tests to use
+	sharedResourceGroup := testContext.NewTestResourceGroup()
+	err = testContext.KubeClient.Create(ctx, sharedResourceGroup)
+	if err != nil {
+		return ControllerTestContext{}, errors.Wrapf(err, "creating shared resource group")
+	}
+
+	// TODO: Should use AzureName here once it's always set
+	log.Printf("Created shared resource group %s\n", sharedResourceGroup.Name)
+
+	// It should be created in Kubernetes
+	// TODO: Generic polling helper needed here
+	// g.Eventually(testContext.Ensure.Created(ctx, rg)).Should(BeTrue())
+
+	log.Println("Done with test setup")
+
+	return ControllerTestContext{testContext, sharedResourceGroup}, nil
+}
+
+func teardown(testContext ControllerTestContext) error {
+	log.Println("Started common controller test teardown")
+
+	ctx := context.Background()
+
+	// List all of the resource groups
+	rgList := &resources.ResourceGroupList{}
+	err := testContext.KubeClient.List(ctx, rgList)
+	if err != nil {
+		return errors.Wrap(err, "listing resource groups")
+	}
+
+	// Delete any leaked resource groups
+	var errs []error
+
+	var resourceGroups []runtime.Object
+
+	for _, rg := range rgList.Items {
+		rg := rg // Copy so that we can safely take addr
+		resourceGroups = append(resourceGroups, &rg)
+		err := testContext.KubeClient.Delete(ctx, &rg)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	err = kerrors.NewAggregate(errs)
+	if err != nil {
+		return err
+	}
+
+	// Don't block forever waiting for delete to complete
+	waitCtx, cancel := context.WithTimeout(ctx, 3*time.Minute)
+	defer cancel()
+
+	allDeleted := false
+
+	for !allDeleted {
+		select {
+		case <-waitCtx.Done():
+			return waitCtx.Err()
+		default:
+			allDeleted, err = testContext.Ensure.AllDeleted(waitCtx, resourceGroups)()
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	log.Println("Finished common controller test teardown")
+	return nil
+}
+
+type controllerTest struct {
+	name string
+	test func(ctx ControllerTestContext, t *testing.T)
+}
+
+func runTests(t *testing.T, tests []controllerTest) {
```

Comment:
Seems like we should be able to make this easier to call than having to create the slice ahead of time.

Reply:
I think I'm going to change things back to using `TestMain` which will allow us to have test autodiscover (but has other issues...)